### PR TITLE
Parameter attributes on Cfunction closures sticks

### LINKED ISSUE #40164: Segfault in ccall
From [discourse](https://discourse.julialang.org/t/why-does-this-ccall-segfault/57690?u=jw3126). The following code segfaults for me:

```julia
C = """
typedef struct CParticle {
    double charge;
    double E;
    double weight;
} CParticle;


void simulate(void (*f)(CParticle x)) {
    CParticle cp;
    cp.charge = 0;
    cp.E = 1;
    cp.weight = 2;
    (*f)(cp);
    return;
}

"""

struct CParticle
    charge::Cdouble
    E::Cdouble
    weight::Cdouble
end

write("lib.c", C)
run(`gcc -c -fPIC lib.c -o lib.o`)
run(`gcc -shared lib.o -o lib.so`)
lib = "./lib.so"

function main()
    ret = CParticle[]
    f = x::CParticle -> (push!(ret, x); nothing)
    f_c = @cfunction $f Cvoid (CParticle,)
    GC.@preserve ret f f_c begin # AFAICT I don't need to GC.@preserver here
        @ccall lib.simulate(f_c::Ptr{Cvoid})::Cvoid
    end
    @show ret
end

main()

```
I read the [manual](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Calling-C-and-Fortran-Code) and in particular the section [about closures](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Closure-cfunctions), but I am still not sure where the error lies.
 It works, if I do one of the following:
* make the `CParticle struct` smaller e.g. only `charge + weight` fields 
* execute the `main` code globally
* don't call `push!` in the callback

@Gnimuc pointed out workarounds, but it is still unclear why this happens, and would be nice to fix it or document it. I can help with the latter.

### PR DESCRIPTION
**Pull request by troels**
When CFunction closures are created an extra parameter is added to the
function signature for holding the closure.

Make sure that the parameter attributes on already existing parameters
are not shifted when adding that parameter.

Fixes  #40164

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -5169,9 +5169,23 @@ static Function* gen_cfun_wrapper(
         // add nest parameter (pointer to jl_value_t* data array) after sret arg
         assert(closure_types);
         std::vector<Type*> fargt_sig(sig.fargt_sig);
-        fargt_sig.insert(fargt_sig.begin() + sig.sret, T_pprjlvalue);
-        functype = FunctionType::get(sig.sret ? T_void : sig.prt, fargt_sig, /*isVa*/false);
-        attributes = attributes.addAttribute(jl_LLVMContext, 1 + sig.sret, Attribute::Nest);
+
+	fargt_sig.insert(fargt_sig.begin() + sig.sret, T_pprjlvalue);
```
Comment (vchuravy) [inline src/codegen.cpp:5173]: Please match the alignment above

Comment (troels) [reply src/codegen.cpp:5173]: Hmmm... Appears that my editor was set to indent using tabs. Fixed now.

**Review by vtjnash — APPROVED** [review]
Seems like this will work. Seems inefficient, but valid.

**Comment by JeffBezanson** [comment]
Welcome, and thanks! Great to have this fixed.

**Comment by troels** [comment]
Thanks!

I did a check of the LLVM code of AttributeList and discovered I made a miss in the number of AttributeSets preceding the parameter attributes. I had mistakenly believed that only return value (Index 0) precedes parameters, but apparently function attributes do too (they are at index UNSIGNED_INT_MAX, just preceding 0). 

I have done a new push with the correct code, keeping both function attributes and return value attributes in their proper place.

File: src/codegen.cpp [hunk 2]
```diff
@@ -5169,7 +5169,24 @@ static Function* gen_cfun_wrapper(
...
+        if (attributes.getNumAttrSets() > static_cast<unsigned>(2 + sig.sret)) {
+            AttrBuilder toShift;
+            // Skip past function and return and sret attributes to the first real parameter
+            for (auto it = attributes.index_begin() + 2 + sig.sret; it != attributes.index_end(); ++it) {
+                AttrBuilder toShiftTemp(attributes.getAttributes(it));
+                attributes = attributes.removeAttributes(jl_LLVMContext, it);
+                attributes = attributes.addAttributes(jl_LLVMContext, it, toShift);
+                toShift = std::move(toShiftTemp);
+            }
+            attributes = attributes.addAttributes(jl_LLVMContext, attributes.index_end(), toShift);
```
Comment (vtjnash) [inline src/codegen.cpp:5187]:
```suggestion
                attributes = attributes.setAttributes(jl_LLVMContext, it, toShift);
                toShift = std::move(toShiftTemp);
            }
            attributes = attributes.setAttributes(jl_LLVMContext, attributes.index_end(), toShift);
```

File: src/codegen.cpp [hunk 3]
```diff
@@ -5169,7 +5169,23 @@ static Function* gen_cfun_wrapper(
...
+        // attributes before the parameter attributes.
+        if (attributes.getNumAttrSets() > static_cast<unsigned>(2 + sig.sret)) {
+            AttrBuilder toShift;
+            // Skip past function and return and sret attributes to the first real parameter
+            for (auto it = attributes.index_begin() + 2 + sig.sret; it != attributes.index_end(); ++it) {
+                AttrBuilder toShiftTemp(attributes.getAttributes(it));
+                attributes = attributes.setAttributes(jl_LLVMContext, it, toShift);
+                toShift = std::move(toShiftTemp);
+            }
+            attributes = attributes.setAttributes(jl_LLVMContext, attributes.index_end(), toShift);
```
Comment (vtjnash) [inline src/codegen.cpp:5186]: oh, oops, this has the wrong type now
```suggestion
            AttributeSet toShift;
            // Skip past function and return and sret attributes to the first real parameter
            for (auto it = attributes.index_begin() + 2 + sig.sret; it != attributes.index_end(); ++it) {
                AttributeSet toShiftTemp(attributes.getAttributes(it));
                attributes = attributes.setAttributes(jl_LLVMContext, it, toShift);
                toShift = toShiftTemp;
            }
            attributes = attributes.setAttributes(jl_LLVMContext, attributes.index_end(), toShift);
```

Comment (troels) [reply src/codegen.cpp]: Whoops, and I noticed now, we of course also need to remove the attributes form the first parameter. I made another version which I think is efficient enough.

Comment (vtjnash) [reply src/codegen.cpp]: I think that this did–the first was set to `toShift = AttributeSet()`

**Review by vtjnash — APPROVED** [review]
Still LGTM to merge (and now more efficient!)

File: src/codegen.cpp [hunk 4]
```diff
@@ -5180,9 +5180,42 @@ static Function* gen_cfun_wrapper(
...
+        auto it = attributes.index_begin();
+
+        // Skip past FunctionIndex
+        if (it == AttributeList::AttrIndex::FunctionIndex) {
+            ++it;
+        }
+
+        // Move past ReturnValue and parameter return value
+        for (;it < AttributeList::AttrIndex::FirstArgIndex + sig.sret; ++it) {
+            newAttributes.emplace_back(it, attributes.getAttributes(it));
```
Comment (vtjnash) [inline src/codegen.cpp:5199]:
```suggestion
            if (attributes.hasAttributes(it))
                newAttributes.emplace_back(it, attributes.getAttributes(it));
```

File: src/codegen.cpp [hunk 5]
```diff
@@ -5180,9 +5180,42 @@ static Function* gen_cfun_wrapper(
...
+        }
+
+        // Add the new nest attribute
+        AttrBuilder attrBuilder;
+        attrBuilder.addAttribute(Attribute::Nest);
+        newAttributes.emplace_back(it, AttributeSet::get(jl_LLVMContext, attrBuilder));
+
+        // Shift forward the rest of the attributes
+        for(;it < attributes.index_end(); ++it) {
+            newAttributes.emplace_back(it + 1, attributes.getAttributes(it));
```
Comment (vtjnash) [inline src/codegen.cpp:5209]:
```suggestion
            if (attributes.hasAttributes(it))
                newAttributes.emplace_back(it + 1, attributes.getAttributes(it));
```

File: src/codegen.cpp [hunk 6]
```diff
@@ -5180,9 +5180,42 @@ static Function* gen_cfun_wrapper(
...
+        }
+
+        // Remember to add back FunctionIndex
+        newAttributes.emplace_back(AttributeList::AttrIndex::FunctionIndex,
+                                   attributes.getAttributes(AttributeList::AttrIndex::FunctionIndex));
```
Comment (vtjnash) [inline src/codegen.cpp:5214]:
```suggestion
        if (attributes.hasFnAttrs())
            newAttributes.emplace_back(AttributeList::AttrIndex::FunctionIndex,
                                       attributes.getFnAttrs());
```

**Comment by vtjnash** [comment]
Looks like this constructor requires dropping !hasAttributes ones (or use the get constructor with separate Fn, Ret, and Params attributes, which permits empty attributes)

**Pull request merged**